core/src/model/fact.rs

```diff
@@ -286,6 +301,15 @@ impl TypedFact {
     pub fn without_value(&self) -> Self {
         Self::dt_shape(self.datum_type, self.shape.clone())
     }
+
+    pub fn opaque_metadata<T: OpaqueMetadata>(&self) -> Option<&T> {
```

Comment:
I'm not too sure about this one, because there is no way to distinguish the case where metadata is missing, and the case where metadata as the wrong type.

It can be written `typed_fact.opaque_metadata.and_then(|om| om.downcast_ref::<...>)`, and it's not much longer...

WDYT ?

Reply:
yes I am changing this